### PR TITLE
IGNITE-16502 fixed deadlock caused by query plan cache

### PR DESCRIPTION
**Pull request by korlov42**

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/QueryPlanCacheImpl.java [hunk 1]
```diff
@@ -18,50 +18,99 @@
...
+
+        return planFut.join().copy();
     }
 
     /** {@inheritDoc} */
     @Override
     public QueryPlan queryPlan(CacheKey key) {
-        Map<CacheKey, QueryPlan> cache = this.cache;
-        QueryPlan plan = cache.get(key);
+        lock.readLock().lock();
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/QueryPlanCacheImpl.java:87]: why not write lock?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/QueryPlanCacheImpl.java [hunk 2]
```diff
@@ -18,50 +18,99 @@
...
+                .<CacheKey, CompletableFuture<QueryPlan>>build()
                 .asMap();
     }
 
     /** {@inheritDoc} */
     @Override
     public QueryPlan queryPlan(CacheKey key, Supplier<QueryPlan> planSupplier) {
-        Map<CacheKey, QueryPlan> cache = this.cache;
-        QueryPlan plan = cache.computeIfAbsent(key, k -> planSupplier.get());
+        lock.readLock().lock();
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/QueryPlanCacheImpl.java:72]: why not write lock?

Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/QueryPlanCacheImpl.java:72]: Let's add comment why we have such locks for cache, it's not obviously now

Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/QueryPlanCacheImpl.java]: I wanted to make clearing the cache and task queue an atomic operation, but in fact we should not clean the task queue. So I removed this lock

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/QueryPlanCacheImpl.java [hunk 3]
```diff
@@ -18,44 +18,93 @@
...
     /** {@inheritDoc} */
     @Override
     public QueryPlan queryPlan(CacheKey key, Supplier<QueryPlan> planSupplier) {
-        Map<CacheKey, QueryPlan> cache = this.cache;
-        QueryPlan plan = cache.computeIfAbsent(key, k -> planSupplier.get());
+        CompletableFuture<QueryPlan> planFut = cache.computeIfAbsent(key, k -> CompletableFuture.supplyAsync(planSupplier, planningPool));
+
+        try {
+            return planFut.join().copy();
+        } catch (CancellationException | CompletionException ex) {
```
Comment (tledkov-gridgain) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/QueryPlanCacheImpl.java:73]: Duplicate conversion exception code. 
`CompletableFuture` is used a lot at the current project. May be we introduce `ExceptionsUtils` (aka X class) to convert the most used exception types?

**Review by tledkov-gridgain — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/QueryPlanCacheImpl.java (hunk 3 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/QueryPlanCacheImpl.java]: Deduplicate the exception handling. But generic ExceptionUtils is outside of the scope of this issue, I believe.

**Review by tledkov-gridgain — APPROVED** [review]

**Pull request merged**